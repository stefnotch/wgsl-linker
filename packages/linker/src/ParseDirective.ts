import {
  anyThrough,
  kind,
  opt,
  or,
  Parser,
  repeat,
  req,
  seq,
  setTraceNames,
  tokens,
  tracing,
  withSep,
} from "mini-parse";
import { gleamImport } from "./GleamImport.js";
import { ImportTree, SimpleSegment } from "./ImportTree.js";
import {
  argsTokens,
  lineCommentTokens,
  mainTokens,
  moduleTokens,
} from "./MatchWgslD.js";
import { eolf, makeElem } from "./ParseSupport.js";

/* parse #directive enhancements to wgsl: #import, #export, etc. */

const argsWord = kind(argsTokens.arg);
const fromWord = or(argsWord, kind(argsTokens.relPath));

// prettier-ignore
/** ( <a> <,b>* ) */
export const directiveArgs: Parser<string[]> = 
  seq(
    "(", 
    withSep(",", argsWord), 
    req(")")
  ).map((r) => r.value[1]);

const fromClause = seq(
  "from",
  or(fromWord.tag("from"), seq('"', fromWord.tag("from"), '"')),
);

export interface ImportClause {
  name: string;
  as?: string;
  args?: string[];
}

// prettier-ignore
/** foo <(A,B)> <as bar> */
const importClause = seq(
  argsWord.tag("name"),
  opt(directiveArgs.tag("args")),
  opt(seq("as", argsWord.tag("as")))
).map(r =>
  ({ name: r.tags.name[0],
    as: r.tags.as?.[0],
    args: r.tags.args?.[0],
  }) as ImportClause
).tag("importClause");

const importList = withSep(",", importClause, { requireOne: true });

// prettier-ignore
const bracketedImportClause = or(
  importList, 
  seq("{", importList, "}")
);

const importElemPhrase = seq(bracketedImportClause, fromClause).map(r => {
  const from = r.tags.from?.[0];
  return r.tags.importClause.map(impClause => {
    const elem = makeElem("treeImport", r as any, [], []);
    const fromSegments = from.split("/").map(s => new SimpleSegment(s));
    const lastSegment = new SimpleSegment(
      impClause.name,
      impClause.as,
      impClause.args,
    );
    const segments = [...fromSegments, lastSegment];
    const importTree: ImportTree = new ImportTree(segments);
    elem.imports = importTree;
    // TODO wildcards
    return elem;
  });
});

if (tracing) setTraceNames({ importElemPhrase });

/** #import foo <(a,b)> <as boo> <from bar>  EOL */
const importDirective = seq(
  or("#import", "import"),
  seq(importElemPhrase.tag("imp"), opt(";"), () => eolf),
).map(r => {
  r.tags.imp[0].forEach(imp => {
    imp.start = r.start; // use start of #import, not import phrase
    r.app.state.push(imp);
  });
});


export const importing = seq(
  "importing",
  seq(importElemPhrase.tag("importing")),
  repeat(seq(",", importElemPhrase.tag("importing"))),
);

/** #export <foo> <(a,b)> <importing bar(a) <zap(b)>* > EOL */
export const exportDirective = seq(
  or("#export", "export"),
  seq(opt(directiveArgs.tag("args")), opt(importing), opt(eolf)),
).map(r => {
  const e = makeElem("export", r, ["args", "importing"]);
  r.app.state.push(e);
});

const moduleDirective = seq(
  or("module", "#module"),
  tokens(moduleTokens, req(kind(moduleTokens.moduleName).tag("name"))),
  eolf,
).map(r => {
  const e = makeElem("module", r);
  e.name = normalizeModulePath(r.tags.name[0]);
  r.app.state.push(e);
});

function normalizeModulePath(name: string): string {
  if (name.includes("::")) {
    const result = name.split("::").join("/");
    return result;
  }
  return name;
}

export const directive = tokens(
  argsTokens,
  seq(
    repeat("\n"),
    or(
      exportDirective,
      importDirective,
      gleamImport,
      moduleDirective,
    ),
  ),
);

const skipToEol = tokens(lineCommentTokens, anyThrough(eolf));

/** parse a line comment */
export const lineComment = seq(tokens(mainTokens, "//"), skipToEol);

if (tracing) {
  setTraceNames({
    directiveArgs,
    fromClause,
    importClause,
    importList,
    bracketedImportClause,
    importElemPhrase,
    importing,
    importDirective,
    exportDirective,
    skipToEol,
    lineComment,
    moduleDirective,
    directive,
  });
}

function copyDefinedProps<S extends Record<string, any>>(
  src: S,
  keys: (keyof S)[],
  dest: any,
): void {
  keys.forEach(k => {
    if (src[k] !== undefined) dest[k] = src[k];
  });
}
