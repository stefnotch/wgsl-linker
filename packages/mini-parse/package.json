{
  "name": "mini-parse",
  "version": "0.5.0-d1",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "pr": "printenv",
    "prepublishOnly": "run-s build",
    "build": "run-s build:main build:util build:minified",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:minified": "vite build --minify --emptyOutDir=false",
    "test": "vitest",
    "organize": "organize-imports-cli tsconfig.json",
    "format": "prettier-eslint \"src/**/*.ts\" --write",
    "lint": "eslint src"
  },
  "type": "module",
  "repository": "github:wgsl-tooling-wg/wgsl-linker",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./minified": {
      "import": "./dist/minified.js",
      "types": "./dist/index.d.ts"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^22.5.0",
    "@typescript-eslint/eslint-plugin": "^8.12.2",
    "@typescript-eslint/parser": "^8.12.2",
    "@vitest/ui": "^2.1.4",
    "berry-pretty": "^0.0.5",
    "eslint": "^9.13.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "typescript": "^5.6.3",
    "vite": "^5.4.10",
    "vite-plugin-dts": "^4.3.0",
    "vite-tsconfig-paths": "^5.0.1",
    "vitest": "^2.1.4"
  }
}