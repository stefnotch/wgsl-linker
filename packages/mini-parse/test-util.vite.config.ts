/// <reference types="vitest" />
import assert from "node:assert";
import { resolve } from "path";
import { LibraryOptions, defineConfig } from "vite";
import baseViteConfig from "./vite.config.js";

const config = baseViteConfig;
assert(config.build);
config.build.emptyOutDir = false;
const lib = config.build.lib as LibraryOptions;
lib.name = "mini-parse-test-util";
lib.formats = ["es"];
lib.entry = [resolve(__dirname, "./src/test-util/index.ts")];
lib.fileName = "testUtil";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);

// {
//   plugins: [
//     tsconfigPaths(),
//     dts(), // generate
//     // visualizer({ brotliSize: true, gzipSize: true }), // generate stats.html size report
//   ],
//   build: {
//     emptyOutDir: false,
//     lib: {
//       name: "mini-parse-testing",
//       formats: ["es"],
//       entry: [resolve(__dirname, "./src/test-util/index.ts")],
//       fileName: "testUtil"
//     },
//     rollupOptions: {
//       // make sure to externalize deps that shouldn't be bundled
//       // into your library
//       external: ['vitest', 'mini-parse'],
//     },
//     minify: false,
//     sourcemap: true,
//   },
// });
