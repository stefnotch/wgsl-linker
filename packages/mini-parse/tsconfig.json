{
  "extends": ["../../tsconfig.options.json"],
  "compilerOptions": {
    "lib": ["DOM", "DOM.Iterable", "ESNext", "ESNext.AsyncIterable"],
    "paths": {
      "mini-parse": ["./src"],
      "mini-parse/test-util": ["./src/test-util"]
    }
  },
  "include": ["src/**/*", "*.config.ts"],
  "exclude": ["node_modules"]
}
