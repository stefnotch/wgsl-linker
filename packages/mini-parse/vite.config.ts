/// <reference types="vitest" />
import { resolve } from "path";
import { defineConfig } from "vite";
import dts from "vite-plugin-dts";

export default defineConfig({
  plugins: [dts()],
  test: { setupFiles: "./src/test/TestSetup.ts" },
  build: {
    lib: {
      name: "mini-parse",
      entry: [resolve(__dirname, "src/index.ts")],
    },
    minify: false,
    sourcemap: true,
  },
});
