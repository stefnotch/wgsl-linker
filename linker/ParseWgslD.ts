import {
  anyNot,
  anyThrough,
  eof,
  ExtendedResult,
  kind,
  matchingLexer,
  opt,
  or,
  Parser,
  ParserContext,
  ParserInit,
  preParse,
  repeat,
  repeatPlus,
  req,
  seq,
  setTraceName,
  simpleParser,
  SrcMap,
  tokens,
  tracing,
  withSep,
} from "@wesl/mini-parse";
import { AbstractElem, TypeNameElem, TypeRefElem } from "./AbstractElems.ts";
import { identTokens, mainTokens } from "./MatchWgslD.ts";
import { directive } from "./ParseDirective.ts";
import {
  comment,
  literal,
  makeElem,
  unknown,
  word,
  wordNumArgs,
} from "./ParseSupport.ts";

/** parser that recognizes key parts of WGSL and also directives like #import */

const longIdent = kind(identTokens.longIdent);

// prettier gets confused if we leave the quoted parens inline so make consts for them here
const lParen = "(";
const rParen = ")";

export interface ParseState {
  params: Record<string, any>; // user provided params to templates, code gen and #if directives
}

const optAttributes = repeat(seq(kind(mainTokens.attr), opt(wordNumArgs)));
const possibleTypeRef = Symbol("typeRef");

const globalDirectiveOrAssert = seq(
  or("diagnostic", "enable", "requires", "const_assert"),
  req(anyThrough(";")),
).map(r => {
  const e = makeElem("globalDirective", r);
  r.app.state.push(e);
});

/** parse an identifier into a TypeNameElem */
export const typeNameDecl = req(word.tag("name")).map(r => {
  return makeElem("typeName", r, ["name"]) as TypeNameElem; // fix?
});

/** parse an identifier into a TypeNameElem */
export const fnNameDecl = req(word.tag("name"), "missing fn name").map(r => {
  return makeElem("fnName", r, ["name"]);
});

/** find possible references to user types (structs) in this possibly nested template */
export const template: Parser<any> = seq(
  "<",
  or(
    word.tag(possibleTypeRef), // only the first element of the template can be a type
    () => template,
  ),
  repeat(
    or(
      () => template,
      anyNot(">"), // we don't care about the rest of the template
    ),
  ),
  req(">"),
);

/** find possible references to user structs in this type specifier and any templates */
export const typeSpecifier: Parser<TypeRefElem[]> = seq(
  tokens(identTokens, longIdent.tag(possibleTypeRef)),
  opt(template),
).map(r =>
  r.tags[possibleTypeRef].map(name => {
    const e = makeElem("typeRef", r as ExtendedResult<any>);
    e.name = name;
    return e as Required<typeof e>;
  }),
);

export const structMember = seq(
  optAttributes,
  word.tag("name"),
  ":",
  req(typeSpecifier.tag("typeRefs")),
).map(r => {
  return makeElem("member", r, ["name", "typeRefs"]);
});

export const structDecl = seq(
  "struct",
  req(typeNameDecl).tag("nameElem"),
  req("{"),
  withSep(",", structMember).tag("members"),
  req("}"),
).map(r => {
  const e = makeElem("struct", r, ["members"]);
  const nameElem = r.tags.nameElem[0];
  e.nameElem = nameElem;
  e.name = nameElem.name;
  r.app.state.push(e);
});

// keywords that can be followed by (), not to be confused with fn calls
const callishKeyword = simpleParser("keyword", (ctx: ParserContext) => {
  const keywords = ["if", "for", "while", "const_assert", "return"];
  const token = ctx.lexer.next();
  const text = token?.text;
  if (text && keywords.includes(text)) {
    return text;
  }
});

export const fnCall = tokens(
  identTokens,
  seq(
    longIdent
      .tag("name")
      .map(r => makeElem("call", r, ["name"]))
      .tag("calls"), // we collect this in fnDecl, to attach to FnElem
    "(",
  ),
);

// prettier-ignore
const fnParam = seq(
  optAttributes,
  word,
  opt(seq(":", req(typeSpecifier.tag("typeRefs"))))
);

const fnParamList = seq(lParen, withSep(",", fnParam), rParen);

// prettier-ignore
const variableDecl = seq(
  or("const", "var", "let", "override"), 
  word, 
  ":", 
  req(typeSpecifier).tag("typeRefs")
);

// TODO: Fix everything else to also use this instead of "template"
const opt_template_args = opt(
  seq(
    "<",
    withSep(",", () => template_arg_expression, {
      requireOne: true,
    }),
    ">",
  ).tag("template"),
);

const primary_expression = or(
  literal,
  seq(
    word,
    opt_template_args,
    opt(
      seq(
        "(",
        withSep(",", () => expression),
        req(")"),
      ),
    ),
  ),
  seq("(", () => expression, req(")")),
);
const component_or_swizzle = repeatPlus(
  or(
    seq(".", word),
    seq("[", () => expression, req("]")),
  ),
);

/**
 * bitwise_expression.post.unary_expression
 * & ^ |
 * expression
 * && ||
 * relational_expression.post.unary_expression
 * > >= < <= != ==
 * shift_expression.post.unary_expression
 * % * / + - << >>
 */
const makeExpressionOperator = (isTemplate: boolean) => {
  const allowedOps = (
    "& | ^ << <= < != == % * / + -" + (isTemplate ? "" : " && || >> >= >")
  ).split(" ");
  return or(...allowedOps)
    .traceName("operator")
    .trace({
      shallow: true,
    });
};
const unary_expression: Parser<any> = or(
  seq(
    or(..."! & * - ~".split(" "))
      .traceName("unary_op")
      .trace({
        shallow: true,
      }),
    () => unary_expression,
  ),
  seq(primary_expression, opt(component_or_swizzle)),
);
const makeExpression = (isTemplate: boolean) => {
  return seq(
    unary_expression,
    repeat(seq(makeExpressionOperator(isTemplate), unary_expression)),
  );
};

export const expression = makeExpression(false);
const template_arg_expression = makeExpression(true);

const statement = repeatPlus(anyNot(or("{", "}"))); // TBD

const compound_statement = seq(optAttributes, "{", repeat(statement), "}");

const case_selector = or("default", expression);
const case_selectors = withSep(",", case_selector, { requireOne: true });
const case_clause = seq("case", case_selectors, opt(":"), compound_statement);
const default_alone_clause = seq("default", opt(":"), compound_statement);
const switch_clause = or(case_clause, default_alone_clause);

const switch_body = seq(optAttributes, "{", repeatPlus(switch_clause), "}");

const switch_statement = seq(optAttributes, "switch", expression, switch_body);

// prettier-ignore
const block: Parser<any> = seq(
  "{",
  repeat(
    or(
      switch_statement,
      callishKeyword,
      fnCall,
      () => block,
      variableDecl,
      anyNot("}")
    )
  ),
  req("}")
);

export const fnDecl = seq(
  optAttributes,
  "fn",
  req(fnNameDecl).tag("nameElem"),
  req(fnParamList),
  opt(seq("->", optAttributes, typeSpecifier.tag("typeRefs"))),
  req(block),
).map(r => {
  const e = makeElem("fn", r);
  const nameElem = r.tags.nameElem[0];
  e.nameElem = nameElem as Required<typeof nameElem>;
  e.name = nameElem.name;
  e.calls = r.tags.calls || [];
  e.typeRefs = r.tags.typeRefs?.flat() || [];
  r.app.state.push(e);
});

export const globalVar = seq(
  optAttributes,
  or("const", "override", "var"),
  opt(template),
  word.tag("name"),
  opt(seq(":", req(typeSpecifier.tag("typeRefs")))),
  req(anyThrough(";")),
).map(r => {
  const e = makeElem("var", r, ["name"]);
  e.typeRefs = r.tags.typeRefs?.flat() || [];
  r.app.state.push(e);
});

export const globalAlias = seq(
  "alias",
  req(word.tag("name")),
  req("="),
  req(typeSpecifier).tag("typeRefs"),
  req(";"),
).map(r => {
  const e = makeElem("alias", r, ["name", "typeRefs"]);
  r.app.state.push(e);
});

const globalDecl = or(fnDecl, globalVar, globalAlias, structDecl, ";");

const rootDecl = or(globalDirectiveOrAssert, globalDecl, directive, unknown);

const root = preParse(comment, seq(repeat(rootDecl), eof()));

export function parseWgslD(
  src: string,
  srcMap?: SrcMap,
  params: Record<string, any> = {},
  maxParseCount: number | undefined = undefined,
  grammar = root,
): AbstractElem[] {
  const lexer = matchingLexer(src, mainTokens);
  const state: AbstractElem[] = [];
  const context: ParseState = { params };
  const app = {
    context,
    state,
  };
  const init: ParserInit = {
    lexer,
    app,
    srcMap,
    maxParseCount,
  };

  grammar.parse(init);

  return app.state;
}

if (tracing) {
  const names: Record<string, Parser<unknown>> = {
    globalDirectiveOrAssert,
    template,
    typeSpecifier,
    structMember,
    structDecl,
    fnCall,
    fnParam,
    fnParamList,
    opt_template_args,
    primary_expression,
    component_or_swizzle,
    expression,
    statement,
    compound_statement,
    case_selector,
    case_selectors,
    case_clause,
    default_alone_clause,
    switch_clause,
    switch_body,
    switch_statement,
    block,
    fnDecl,
    globalVar,
    globalAlias,
    globalDecl,
    rootDecl,
    root,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
