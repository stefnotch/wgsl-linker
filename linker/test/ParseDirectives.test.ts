import { _withBaseLogger, tokens } from "@wesl/mini-parse";
import { logCatch } from "@wesl/mini-parse/test-util";

import { expect, test } from "vitest";
import { assertSnapshot } from "@std/testing/snapshot";
import { ModuleElem, TreeImportElem } from "../AbstractElems.ts";
import { SimpleSegment, treeToString } from "../ImportTree.ts";
import { argsTokens } from "../MatchWgslD.ts";
import { directive, importing } from "../ParseDirective.ts";
import { parseWgslD } from "../ParseWgslD.ts";
import { last } from "../Util.ts";
import { testAppParse } from "./TestUtil.ts";

test("directive parses #export", () => {
  const { appState } = testAppParse(directive, "#export");
  expect(appState[0].kind).toBe("export");
});

test("parse #export", () => {
  const parsed = parseWgslD("#export");
  expect(parsed[0].kind).toBe("export");
});

test("parse import foo/bar", async (ctx) => {
  const parsed = parseWgslD("import foo/bar");
  await assertSnapshot(ctx, parsed);
});

test("parse #import foo(a,b) as baz from bar", async (ctx) => {
  const parsed = parseWgslD("#import foo as baz from bar");
  await assertSnapshot(ctx, parsed);
});

test("parse #export(foo) with trailing space", async (ctx) => {
  const src = `
    export (Elem) 
  `;

  const parsed = parseWgslD(src);
  await assertSnapshot(ctx, parsed);
});

test.ignore("importing parses importing bar(A) fog(B)", async (ctx) => {
  const src = ` importing bar(A), fog(B)`;
  const { parsed } = testAppParse(tokens(argsTokens, importing), src);
  await assertSnapshot(ctx, parsed?.tags.importing);
});

test.ignore("parse #export(A, B) importing bar(A)", async (ctx) => {
  const src = `
    #export(A, B) importing bar(A)
    fn foo(a:A, b:B) { bar(a); }
  `;
  const parsed = parseWgslD(src);
  await assertSnapshot(ctx, parsed[0]);
});

test("#export w/o closing paren", async (ctx) => {
  const src = `#export (A
    )
    `;
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => parseWgslD(src));
  await assertSnapshot(ctx, logged());
});

test.ignore("parse #extends", async (ctx) => {
  const src = `#extends Foo(a,b) as Bar from baz`;
  const appState = parseWgslD(src);
  await assertSnapshot(ctx, appState[0]);
});

test("parse extends", async (ctx) => {
  const src = `extends Foo(a,b) as Bar from baz`;
  const appState = parseWgslD(src);
  await assertSnapshot(ctx, appState[0]);
});

test("parse module foo.bar.ca", () => {
  const src = `module foo.bar.ca`;
  const appState = parseWgslD(src);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo.bar.ca");
});

test("module foo.bar.ca", (ctx) => {
  const appState = parseWgslD(ctx.name);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo.bar.ca");
});

test("module foo::bar::ba", (ctx) => {
  const appState = parseWgslD(ctx.name);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo/bar/ba");
});

test("module foo/bar/ba", (ctx) => {
  const appState = parseWgslD(ctx.name);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo/bar/ba");
});

test("parse import with numeric types", () => {
  const nums = "1u 2.0F 0x010 -7.0 1e7".split(" ");
  const src = `#import foo(${nums.join(",")}) from bar`;
  const appState = parseWgslD(src);

  const segments = (appState[0] as TreeImportElem).imports.segments;
  const lastSegment = last(segments) as SimpleSegment;
  expect(lastSegment.args).toEqual(nums);
});

test("#import foo from ./util", (ctx) => {
  const appState = parseWgslD(ctx.name);
  const importElem = appState[0] as TreeImportElem;
  const segments = treeToString(importElem.imports);
  expect(segments).toBe("./util/foo");
});

test('import { foo } from "./bar"', (ctx) => {
  const appState = parseWgslD(ctx.name);
  const importElem = appState[0] as TreeImportElem;
  const segments = treeToString(importElem.imports);
  expect(segments).toBe("./bar/foo");
});

test('import { foo, bar } from "./bar"', (ctx) => {
  const appState = parseWgslD(ctx.name);
  const imports = appState.filter((e) => e.kind === "treeImport");
  const segments = imports.map((i) => treeToString(i.imports));
  expect(segments).toContain("./bar/foo");
  expect(segments).toContain("./bar/bar");
});
