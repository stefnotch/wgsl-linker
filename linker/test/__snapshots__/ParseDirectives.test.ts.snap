export const snapshot = {};

snapshot[`parse import foo/bar 1`] = `
[
  {
    end: 14,
    imports: ImportTree {
      segments: [
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: "foo",
        },
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: "bar",
        },
      ],
    },
    kind: "treeImport",
    start: 0,
  },
]
`;

snapshot[`parse #import foo(a,b) as baz from bar 1`] = `
[
  {
    end: 27,
    imports: ImportTree {
      segments: [
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: "bar",
        },
        SimpleSegment {
          args: undefined,
          as: "baz",
          name: "foo",
        },
      ],
    },
    kind: "treeImport",
    start: 0,
  },
]
`;

snapshot[`parse #export(foo) with trailing space 1`] = `
[
  {
    args: [
      "Elem",
    ],
    end: 20,
    kind: "export",
    start: 5,
  },
]
`;

snapshot[`#export w/o closing paren 1`] = `
"expected text ')''
#export (A   Ln 1
          ^"
`;

snapshot[`parse extends 1`] = `
{
  args: [
    "a",
    "b",
  ],
  as: "Bar",
  end: 32,
  from: "baz",
  kind: "extends",
  name: "Foo",
  start: 0,
}
`;
